src/gtest/test_foundersreward.cpp

```diff
@@ -0,0 +1,173 @@
+        s += "    ";
+        for (int col=0; col<colsPerRow; col++) {
+            s += "\"" + addresses[i++] + "\", ";
+        }
+        s += "\n";
+    }
+    s += "    };";
+    std::cout << s << std::endl;
+}
+#endif
+
+
+// Utility method to check the number of unique addresses from height 1 to maxHeight
+void checkNumberOfUniqueAddresses(int nUnique) {
+    int maxHeight = Params().GetConsensus().GetLastFoundersRewardBlockHeight();
+    std::set<std::string> addresses;
+    for (int i = 1; i <= maxHeight; i++) {
+        addresses.insert(Params().GetFoundersRewardAddressAtHeight(i));
+    }
+    ASSERT_TRUE(addresses.size() == nUnique);
+}
+
+
+TEST(founders_reward_test, general) {
+    SelectParams(CBaseChainParams::TESTNET);
+
+    CChainParams params = Params();
+    
+    // For Testnet, the first address is derived from the old script. 
+    // FOUNDERS_REWARD_SCRIPT = a9146708e6670db0b950dac68031025cc5b63213a49187
+    //                address = 2N2e2FRfP9D1dRN1oRWkH7pbFM69eGNAuQ4
+    EXPECT_EQ(params.GetFoundersRewardScriptAtHeight(1), ParseHex("a9146708e6670db0b950dac68031025cc5b63213a49187"));
+    EXPECT_EQ(params.GetFoundersRewardAddressAtHeight(1), "2N2e2FRfP9D1dRN1oRWkH7pbFM69eGNAuQ4");
+
+    int maxHeight = params.GetConsensus().GetLastFoundersRewardBlockHeight();
+    
+    // If the block height parameter is out of bounds, there is an assert.
+    EXPECT_DEATH(params.GetFoundersRewardScriptAtHeight(0), "nHeight");
+    EXPECT_DEATH(params.GetFoundersRewardScriptAtHeight(maxHeight+1), "nHeight");
+    EXPECT_DEATH(params.GetFoundersRewardAddressAtHeight(0), "nHeight");
+    EXPECT_DEATH(params.GetFoundersRewardAddressAtHeight(maxHeight+1), "nHeight"); 
+}
+
+
+#define NUM_MAINNET_FOUNDER_ADDRESSES 48
+
+TEST(founders_reward_test, mainnet) {
+    SelectParams(CBaseChainParams::MAIN);
+    checkNumberOfUniqueAddresses(NUM_MAINNET_FOUNDER_ADDRESSES);
+}
+
+
+#define NUM_TESTNET_FOUNDER_ADDRESSES 48
+
+TEST(founders_reward_test, testnet) {
+    SelectParams(CBaseChainParams::TESTNET);
+    checkNumberOfUniqueAddresses(NUM_TESTNET_FOUNDER_ADDRESSES);
+}
+
+
+#define NUM_REGTEST_FOUNDER_ADDRESSES 1
+
+TEST(founders_reward_test, regtest) {
+    SelectParams(CBaseChainParams::REGTEST);
+    checkNumberOfUniqueAddresses(NUM_REGTEST_FOUNDER_ADDRESSES);
+}
+
+
+
+// Test that 10% founders reward is fully rewarded after the first halving and slow start shift.
+// On Mainnet, this would be 2,100,000 ZEC after 850,000 blocks (840,000 + 10,000).
+TEST(founders_reward_test, slow_start_subsidy) {
+    SelectParams(CBaseChainParams::MAIN);
+    CChainParams params = Params();
+
+    int maxHeight = params.GetConsensus().GetLastFoundersRewardBlockHeight();    
+    CAmount totalSubsidy = 0;
+    for (int nHeight = 1; nHeight <= maxHeight; nHeight++) {
+        CAmount nSubsidy = GetBlockSubsidy(nHeight, params.GetConsensus()) / 5;
+        totalSubsidy += nSubsidy;
+    }
+    
+    ASSERT_TRUE(totalSubsidy == MAX_MONEY/10.0);
```

Comment:
Sanity check: this definitely won't ever do integer division? It might be better for future-proofing to use `totalSubsidy * 10.0 == MAX_MONEY`.